securedrop/qa_loader.py

```diff
@@ -0,0 +1,226 @@
+        os.mkdir(source_dir)
+
+    filename = random_chars(20, nullable=False, chars=string.ascii_lowercase)
+    filename = '1-' + filename + '-msg.gpg'
+    f_len = int(math.floor(random.expovariate(10) * 1024 * 1024 * 500))
+    sub_path = current_app.storage.path(source_fid, filename)
+    with open(sub_path, 'w') as f:
+        f.write('x' * f_len)
+
+    return filename
+
+
+def new_source_star(source_id):
+    source = Source.query.get(source_id)
+    star = SourceStar(source, bool_or_none())
+    db.session.add(star)
+
+
+def new_reply(config, journalist_id, source_id):
+    source = Source.query.get(source_id)
+
+    # A source may have a null fid according to the DB, but this will
+    # break storage.path.
+    if source.filesystem_id is None:
+        return
+
+    journalist = Journalist.query.get(journalist_id)
+    filename = fake_file(config, source.filesystem_id)
+    reply = Reply(journalist, source, filename)
+    db.session.add(reply)
+
+
+def new_journalist_login_attempt(journalist_id):
+    journalist = Journalist.query.get(journalist_id)
+    attempt = JournalistLoginAttempt(journalist)
+    attempt.timestamp = random_datetime(nullable=True)
+    db.session.add(attempt)
+
+
+def new_abandoned_submission(config, source_id):
+    '''For issue #1189'''
+
+    source = Source.query.filter(Source.filesystem_id.isnot(None)).all()[0]
+    filename = fake_file(config, source.filesystem_id)
+
+    # Use this as hack to create a real submission then swap out the source_id
+    submission = Submission(source, filename)
+    submission.source_id = source_id
+    db.session.add(submission)
+
+
+def positive_int(s):
+    i = int(s)
+    if i < 1:
+        raise ValueError('{} is not >= 1'.format(s))
+    return i
+
+
+def load_data(config, multiplier):
+    app = create_app(config)
+
+    with app.app_context():
+        for _ in range(JOURNALIST_COUNT * multiplier):
+            new_journalist()
+        db.session.commit()
+
+        for _ in range(SOURCE_COUNT * multiplier):
+            new_source()
+        db.session.commit()
+
+        for sid in range(1, SOURCE_COUNT * multiplier, 5):
+            for _ in range(1, multiplier + 1):
+                new_submission(config, sid)
+        db.session.commit()
+
+        for sid in range(1, SOURCE_COUNT * multiplier, 5):
+            new_source_star(sid)
+        db.session.commit()
+
+        for jid in range(1, JOURNALIST_COUNT * multiplier, 10):
+            for sid in range(1, SOURCE_COUNT * multiplier, 10):
+                for _ in range(1, multiplier + 1):
+                    new_reply(config, jid, sid)
+
+        for jid in range(1, JOURNALIST_COUNT * multiplier, 10):
+            new_journalist_login_attempt(jid)
+
+        for sid in range(SOURCE_COUNT * multiplier,
+                         SOURCE_COUNT * multiplier + multiplier):
+            new_abandoned_submission(config, sid)
+
+        # TODO add submissions without sources for #1189
```

Comment:
I think we do want this - `new_abandoned_submission` almost handles this situation, it looks like we just need to delete the source(s) that the abandoned submissions are associated with. Using raw SQL to do the deletion should do the trick since the submission cascade deletion is enforced at the ORM level.